src/connector/src/sink/mongodb.rs

```diff
@@ -0,0 +1,860 @@
+        if self.pk_indices.len() > 1
+            || self.schema.fields[self.pk_indices[0]].name != MONGODB_PK_NAME
+        {
+            // compound pk should not have a field named `_id`
+            document.insert(MONGODB_PK_NAME, pk.clone());
+        }
+
+        let pk = doc! {MONGODB_PK_NAME: pk};
+        match op {
+            Op::Insert | Op::UpdateInsert => {
+                match self.upsert_builder.as_mut().unwrap().entry(ns) {
+                    Entry::Occupied(mut entry) => entry.get_mut().add_upsert(pk, document),
+                    Entry::Vacant(entry) => {
+                        let mut builder = UpsertCommandBuilder::new(coll, self.max_entries);
+                        builder.add_upsert(pk, document)?;
+                        entry.insert(builder);
+                        Ok(())
+                    }
+                }
+            }
+            Op::UpdateDelete => Ok(()),
+            Op::Delete => match self.upsert_builder.as_mut().unwrap().entry(ns) {
+                Entry::Occupied(mut entry) => entry.get_mut().add_delete(pk),
+                Entry::Vacant(entry) => {
+                    let mut builder = UpsertCommandBuilder::new(coll, self.max_entries);
+                    builder.add_delete(pk)?;
+                    entry.insert(builder);
+                    Ok(())
+                }
+            },
+        }
+    }
+
+    async fn write<'a>(&'a mut self, op: Op, row: RowRef<'a>) -> Result<()> {
+        if self.is_append_only {
+            if op != Op::Insert {
+                return Ok(());
+            }
+            self.append(row);
+        } else {
+            if op == Op::UpdateDelete {
+                // we should ignore the `UpdateDelete` in upsert mode
+                return Ok(());
+            }
+            self.upsert(op, row)?;
+        }
+
+        self.buffered_entries += 1;
+        if self.buffered_entries >= self.max_entries {
+            self.flush().await?;
+        }
+
+        Ok(())
+    }
+
+    async fn flush(&mut self) -> Result<()> {
+        if self.is_append_only {
+            if let Some(mut insert_builder) = self.insert_builder.take() {
+                for (ns, builder) in insert_builder.drain() {
+                    self.send_bulk_write_command(&ns.0, builder.build()).await?;
+                }
+                self.insert_builder = Some(insert_builder);
+            }
+        } else if let Some(mut upsert_builder) = self.upsert_builder.take() {
+            for (ns, builder) in upsert_builder.drain() {
+                let (upsert, delete) = builder.build();
+                // we are sending the bulk upsert first because, under same pk, the `Insert` and `UpdateInsert`
+                // should always appear before `Delete`. we have already ignored the `UpdateDelete`
+                // which is useless in upsert mode.
+                if upsert.is_some() {
+                    self.send_bulk_write_command(&ns.0, upsert.unwrap()).await?;
```

Comment:
May use `if let Some(upsert) = upsert`?

Besides, instead of sending command one by one (await in for-loop), is it possible to send command concurrently with `try_join_all`?

Reply:
> Besides, instead of sending command one by one (await in for-loop), is it possible to send command concurrently with try_join_all?
Sounds reasonable, but I think we should handle this performance improvement in a separate PR because it involves additional considerations,  such as how many write commands should be sent to mongodb concurrently? Given that when `dynamic collection` is enabled, rows in a single epoch may accumulate thousands of collections need to be sunk. If we send them simultaneously, we may impact the stability of the mongodb, which is not desired by users. I think it's up to user to specify the concurrency, as well as the user should know how to set properties like `maxPoolSize`, `minPoolSize ` and `maxConnecting ` to make it work properly.